src/storage.rs

```diff
@@ -357,32 +432,31 @@ impl Storage for MemStorage {
 
     /// Implements the Storage trait.
     fn first_index(&self) -> Result<u64> {
-        let core = self.rl();
-        Ok(core.entries[0].get_index() + 1)
+        Ok(self.rl().inner_first_index())
     }
 
     /// Implements the Storage trait.
     fn last_index(&self) -> Result<u64> {
-        let core = self.rl();
-        Ok(core.inner_last_index())
+        Ok(self.rl().inner_last_index())
     }
 
     /// Implements the Storage trait.
     fn snapshot(&self) -> Result<Snapshot> {
         let core = self.rl();
-        Ok(core.snapshot.clone())
+        Ok(core.inner_snapshot())
     }
 }
 
 #[cfg(test)]
 mod test {
-    use crate::eraftpb::{ConfState, Entry, Snapshot};
-    use crate::errors::{Error as RaftError, StorageError};
-    use crate::storage::{MemStorage, Storage};
+    use std::panic::{self, AssertUnwindSafe};
+
     use harness::setup_for_test;
     use protobuf;
 
-    // TODO extract these duplicated utility functions for tests
+    use crate::eraftpb::{ConfState, Entry, Snapshot};
+    use crate::errors::{Error as RaftError, StorageError};
+    use crate::storage::{MemStorage, Storage};
```

Comment:
Could use `use super::*;` (which is one of the few places I think using a glob import is OK :-) )